### PR TITLE
Address #13235: avoid passing degenerate in-hyps clauses

### LINKED ISSUE #13235: change ... with ... in empty occurrence clause should be a syntax error
```coq
Goal forall (a: nat), 1 + a < a + 2.
Proof.
  intros.
  change (1 + a) with (S a) in .
```

Actual behavior: The `change` command does nothing.

Desired behavior: The line containing `change` should produce a syntax error.

Motivation: I can't imagine a case where an empty occurrence clause would be a useful feature, but I just encountered a bug after refactoring a proof script that would have been caught much faster if this was a syntax error.

Coq version: 8.12.0

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** enhancement

We do not accept a degenerate `in-hyps` clause, which is somehow fragile, as reported in #13235 

In passing, we update the documentation of `change` which was not mentioning the general form of `in-hyps` clause that `change` was actually supporting.

Closes #13235 

- [X] Corresponding documentation was added / updated
- [X] Entry added in the changelog

I did not put a test which would require being able to observe a parsing error (am I correct that we don't know how to do that?).

### CODE REVIEW

File: doc/sphinx/proof-engine/tactics.rst [hunk 1]
```diff
@@ -2941,9 +2941,13 @@ simply :g:`t=u` dropping the implicit type of :g:`t` and :g:`u`.
       .. exn:: Too few occurrences.
          :undocumented:
 
-   .. tacv:: change @term {? {? at {+ @natural}} with @term} in @ident
+   .. tacv:: change @term {? {? at {+ @natural}} with @term} in @goal_occurrences
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/tactics.rst:2944]: There's no nonterminal names `goal_occurrences`, despite being used in many places in this chapter, which has not been updated.  The current syntax shown here is pretty incorrect.  It's something like this (below) in `fullGrammar`, before editing by doc_grammar.  Also, doc_grammar seems to be incorrectly dropping `conversion` somehow, which I'll need to investigate.

Maybe it's best to wait on updating this part of the doc?

```
simple_tactic: [
| "change" conversion clause_dft_concl

conversion: [
| constr
| constr "with" constr
| constr "at" occs_nums "with" constr

clause_dft_concl: [
| "in" G_TACTIC_in_clause
| occs

G_TACTIC_in_clause: [
| "*" occs
| "*" "|-" concl_occ
| LIST0 hypident_occ SEP "," "|-" concl_occ
| LIST0 hypident_occ SEP ","

File: doc/changelog/04-tactics/13237-master+fix13235-no-degenerate-in-hyps-clause.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+- **Changed:**
+  Prevent the :n:`in @goal_occurrences` clause of tactics to silently be empty
```
Comment (jfehrle) [inline doc/changelog/04-tactics/13237-master+fix13235-no-degenerate-in-hyps-clause.rst:2]:
```suggestion
  Require at least one occurrence in TBD
```

Better to state this positively.  Can't use `goal_occurrences`, there's no such nonterminal.

**Review by jfehrle — CHANGES_REQUESTED** [review]
Needs some refinement.

File: plugins/ltac/g_tactic.mlg [hunk 3]
```diff
@@ -379,10 +379,12 @@ GRAMMAR EXTEND Gram
...
           { {onhyps=None; concl_occs=occs} }
-      | hl=LIST0 hypident_occ SEP","; "|-"; occs=concl_occ ->
+      | "|-"; occs=concl_occ ->
+          { {onhyps=Some []; concl_occs=occs} }
+      | hl = LIST0 hypident_occ SEP ","; "|-"; occs=concl_occ ->
           { {onhyps=Some hl; concl_occs=occs} }
-      | hl=LIST0 hypident_occ SEP"," ->
-          { {onhyps=Some hl; concl_occs=NoOccurrences} } ] ]
+      | hl = LIST0 hypident_occ SEP "," ->
+          { if hl = [] then err () else {onhyps=Some hl; concl_occs=NoOccurrences} } ] ]
```
[the remark this answers is not in the document]
Comment (jfehrle) [reply plugins/ltac/g_tactic.mlg:387]: > > why not make this LIST1 hypident_occ in all of these productions?

> In the case there is a `|-`, we need to have the possibility of no hyps before the `|-`, so a `LIST1` would not work.

I should have said "for this production" (line 386).  Unless the parser can't handle that.  A small detail, just an observation.

**Comment by gares** [comment]
@herbelin this is not in a bad shape, but the doc needs some fixes (refman fails to compile)

**Comment by herbelin** [comment]
@zimmi48, @jfehrle: do you understand why the refman fails with "/builds/coq/coq/doc/sphinx/proofs/writing-proofs/rewriting.rst:279:'token' reference target not found: @goal_occurrences" while `@goal_occurrences` occurs in file `doc/sphinx/proof-engine/tactics.rst`.

**Comment by Zimmi48** [comment]
No, I don't understand.

File: doc/changelog/04-tactics/13237-master+fix13235-no-degenerate-in-hyps-clause.rst (hunk 2 above)
Comment (jfehrle) [reply doc/changelog/04-tactics/13237-master+fix13235-no-degenerate-in-hyps-clause.rst:2]: FYI tactics.rst hasn't been revised.  That definition isn't come from doc_grammar.  FWIW #13343 will update the occurrences grammar.

File: doc/sphinx/proofs/writing-proofs/rewriting.rst [hunk 4]
```diff
@@ -274,9 +274,13 @@ simply :g:`t=u` dropping the implicit type of :g:`t` and :g:`u`.
       .. exn:: Too few occurrences.
          :undocumented:
 
-   .. tacv:: change @term {? {? at {+ @natural}} with @term} in @ident
+   .. tacv:: change @term {? {? at {+ @natural}} with @term} in @goal_occurrences
 
-      This applies the :tacn:`change` tactic not to the goal but to the hypothesis :n:`@ident`.
+      In the presence of :n:`with`, this applies :tacn:`change` to the
+      occurrences specified by :token:`@goal_occurrences`. In the
+      absense of :n:`with`, :token:`@goal_occurrences` is expected to
```
Comment (jfehrle) [inline doc/sphinx/proofs/writing-proofs/rewriting.rst:281]:
```suggestion
      occurrences specified by :token:`goal_occurrences`. In the
      absence of :n:`with`, :token:`goal_occurrences` is expected to
```

This will fix your build.  `:token:` doesn't use `@` unlike `:n:`

File: plugins/ltac/g_tactic.mlg [hunk 5]
```diff
@@ -379,10 +379,12 @@ GRAMMAR EXTEND Gram
...
       | "*"; "|-"; occs=concl_occ ->
           { {onhyps=None; concl_occs=occs} }
-      | hl=LIST0 hypident_occ SEP","; "|-"; occs=concl_occ ->
+      | "|-"; occs=concl_occ ->
+          { {onhyps=Some []; concl_occs=occs} }
+      | hl = LIST0 hypident_occ SEP ","; "|-"; occs=concl_occ ->
           { {onhyps=Some hl; concl_occs=occs} }
-      | hl=LIST0 hypident_occ SEP"," ->
-          { {onhyps=Some hl; concl_occs=NoOccurrences} } ] ]
+      | hl = LIST0 hypident_occ SEP "," ->
```
Comment (jfehrle) [inline plugins/ltac/g_tactic.mlg:386]: Why not LIST1 here, since an empty list gives an error anyway?

File: plugins/ltac/g_tactic.mlg [hunk 6]
```diff
@@ -379,10 +379,12 @@ GRAMMAR EXTEND Gram
           { {onhyps=None; concl_occs=occs} }
       | "*"; "|-"; occs=concl_occ ->
           { {onhyps=None; concl_occs=occs} }
-      | hl=LIST0 hypident_occ SEP","; "|-"; occs=concl_occ ->
+      | "|-"; occs=concl_occ ->
+          { {onhyps=Some []; concl_occs=occs} }
+      | hl = LIST0 hypident_occ SEP ","; "|-"; occs=concl_occ ->
```
Comment (jfehrle) [inline plugins/ltac/g_tactic.mlg:384]: And LIST1 here as well since the previous production handles the empty list?  Though I think splitting the old rule into 2 isn't necessary.  (Hope we're not going to go in circles on this detail.)

**Review by jfehrle — COMMENTED** [review]
There's a fix for the refman build failure below.

File: plugins/ltac/g_tactic.mlg (hunk 6 above)
Comment (herbelin) [reply plugins/ltac/g_tactic.mlg]: Using `LIST0` and `LIST1` would not be factorized (I don't think that caml5 does the factorization itself, even if it could, to the price of a less readable grammar), but it seems that I indeed got confused and mixed two approaches. The previous empty list production should be enough to allow `LIST1` in the next two productions. Retrying...

File: doc/sphinx/proofs/writing-proofs/rewriting.rst (hunk 4 above)
Comment (herbelin) [reply doc/sphinx/proofs/writing-proofs/rewriting.rst]: Thanks for the fix. I use ``:n:`@goal_occurrences` `` then. Should be ok, no?

Comment (jfehrle) [reply doc/sphinx/proofs/writing-proofs/rewriting.rst]: Yup.

**Comment by gares** [comment]
@jfehrle are the changes OK? can we merge this one?

**Comment by jfehrle** [comment]
Just one detail left:
@herbelin Can you improve the wording mentioned here: (https://github.com/coq/coq/pull/13237/files#r510527118)?  You would have to provide wording for the "TBD"?

**Comment by herbelin** [comment]
Maybe "The :n:`in @goal_occurrences` clause of tactics cannot be empty", or "The :n:`in @goal_occurrences` clause of tactics needs to mention at least one occurrence"?

Or are your concerns also about referring to the name `@goal_occurrences`?

**Comment by jfehrle** [comment]
I suggest getting rid of `@goal_occurrences`, which is not a nonterminal in the mlgs.  But the reader probably won't understand what `@in_clause` means.  `in_clause` is used in `clause_dft_concl` and `clause_dft_all`, which provide the "in", but those names aare equally opaque.  Rather than trying to explain all that, I'd avoid the nonterminal name altogether.  Imprecision is sometimes advantageous.

How about this wording: `Giving an empty list of occurrences after "in" in tactics is no longer permitted.  Omitting the "in" gives the same behavior.`?

**Comment by herbelin** [comment]
> Giving an empty list of occurrences after "in" in tactics is no longer permitted. Omitting the "in" gives the same behavior.

OK, committed.

**Review by jfehrle — APPROVED** [review]
Copacetic!

**Comment by gares** [comment]
CI green modulo windows, @ppedrot please review/merge

**Review by ppedrot — APPROVED** [review]
